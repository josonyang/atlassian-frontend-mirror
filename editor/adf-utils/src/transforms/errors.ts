export class NodeNestingTransformError extends Error {
	constructor(message: string) {
		super(message);
		this.name = 'NodeNestingTransformError';
	}
}
