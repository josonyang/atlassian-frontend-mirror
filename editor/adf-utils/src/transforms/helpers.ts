import { type ADFEntity } from '../types';

export const isEmpty = (node: ADFEntity) => !node?.content?.length;
