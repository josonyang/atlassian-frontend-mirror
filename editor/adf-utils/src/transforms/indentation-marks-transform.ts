import { traverse } from '../traverse/traverse';
import { type ADFEntity } from '../types';

const hasChildHeadingWithIndentation = (node: ADFEntity): boolean =>
	node.content?.some(
		(childNode) =>
			childNode?.type === 'heading' &&
			childNode?.marks?.some((mark) => mark.type === 'indentation'),
	) ?? false;

const removeIndentationFromHeadings = (node: ADFEntity) => {
	return {
		...node,
		content: node.content?.map((childNode) => {
			if (childNode?.type === 'heading') {
				return {
					...childNode,
					marks: childNode.marks?.filter((mark) => mark.type !== 'indentation'),
				};
			}
			return childNode;
		}),
	};
};

interface TransformIndentationMarksResult {
	isTransformed: boolean;
	transformedAdf: ADFEntity;
}

export const transformIndentationMarks = (adf: ADFEntity): TransformIndentationMarksResult => {
	let isTransformed: boolean = false;

	const transformedAdf = traverse(adf, {
		tableCell: (node) => {
			if (hasChildHeadingWithIndentation(node)) {
				isTransformed = true;
				return removeIndentationFromHeadings(node);
			}
			return;
		},
		tableHeader: (node) => {
			if (hasChildHeadingWithIndentation(node)) {
				isTransformed = true;
				return removeIndentationFromHeadings(node);
			}
			return;
		},
	}) as ADFEntity;

	return {
		transformedAdf,
		isTransformed,
	};
};
