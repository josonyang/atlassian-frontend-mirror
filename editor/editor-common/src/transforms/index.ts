// Disable no-re-export rule for entry point files
/* eslint-disable @atlaskit/editor/no-re-export */

export {
	unwrapContentFromLayout,
	removeLayoutFromFirstChild,
	removeLayoutFromLastChild,
	transformSliceToRemoveOpenLayoutNodes,
	transformSingleColumnLayout,
} from './layout';
export {
	findExpand,
	transformSliceToRemoveOpenExpand,
	transformSliceToRemoveOpenNestedExpand,
	transformSliceNestedExpandToExpand,
	transformSliceExpandToNestedExpand,
} from './expand';
export {
	transformSliceToRemoveOpenBodiedExtension,
	transformSliceToRemoveOpenMultiBodiedExtension,
} from './extension';
export {
	transformSliceToJoinAdjacentCodeBlocks,
	transformSingleLineCodeBlockToCodeMark,
	findCodeBlock,
} from './code-block';
export { transformSliceToDecisionList } from './decision-list';
