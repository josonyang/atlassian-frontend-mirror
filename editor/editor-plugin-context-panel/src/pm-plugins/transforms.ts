import { type Transaction } from '@atlaskit/editor-prosemirror/state';

import { pluginKey } from '../contextPanelPlugin';

export const applyChange = (tr: Transaction): Transaction =>
	tr.setMeta(pluginKey, { changed: true });
