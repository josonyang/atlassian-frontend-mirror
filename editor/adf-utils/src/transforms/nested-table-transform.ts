import type { ADFEntity } from '../types';
import { traverse } from '../traverse/traverse';
import { extension } from '../builders';
import { NodeNestingTransformError } from './errors';

const NESTED_TABLE_EXTENSION_TYPE = 'com.atlassian.nesting',
	NESTED_TABLE_EXTENSION_KEY = 'nested-table';

const isNestedTableExtension = (extensionNode: ADFEntity) =>
	extensionNode.attrs?.extensionType === NESTED_TABLE_EXTENSION_TYPE &&
	extensionNode.attrs?.extensionKey === NESTED_TABLE_EXTENSION_KEY;

const transformNestedTableExtension = (nestedTableExtension: ADFEntity): ADFEntity | false => {
	// No content - drop the extension node
	if (!nestedTableExtension.attrs?.parameters?.macroParams?.nestedContent) {
		return false;
	}

	try {
		return JSON.parse(nestedTableExtension.attrs?.parameters?.macroParams?.nestedContent.value);
	} catch (e) {
		throw new NodeNestingTransformError('Failed to parse nested table content');
	}
};

export const transformNestedTablesIncomingDocument = (
	adf: ADFEntity,
): {
	transformedAdf: ADFEntity;
	isTransformed: boolean;
} => {
	let isTransformed: boolean = false;

	const transformedAdf = traverse(adf, {
		extension: (node) => {
			if (isNestedTableExtension(node)) {
				isTransformed = true;
				return transformNestedTableExtension(node);
			}
		},
	}) as ADFEntity;

	return {
		transformedAdf,
		isTransformed,
	};
};

export const transformNestedTableNodeOutgoingDocument = (tableCellNode: ADFEntity): ADFEntity => {
	try {
		return {
			...tableCellNode,
			content: tableCellNode.content?.map((childNode) => {
				// wrap nested table in an extension node
				if (childNode?.type === 'table') {
					return extension({
						extensionType: NESTED_TABLE_EXTENSION_TYPE,
						extensionKey: NESTED_TABLE_EXTENSION_KEY,
						parameters: {
							macroParams: {
								nestedContent: {
									value: JSON.stringify(childNode),
								},
							},
						},
					});
				}
				return childNode;
			}),
		};
	} catch (e) {
		throw new NodeNestingTransformError('Failed to encode nested table node');
	}
};
