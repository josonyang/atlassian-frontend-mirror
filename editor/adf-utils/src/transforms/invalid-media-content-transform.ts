import { traverse } from '../traverse/traverse';
import { type ADFEntity } from '../types';
import { isEmpty } from './helpers';

const getChildrenTypeCounts: (
	nodeContent: (ADFEntity | undefined)[],
	allowedTypes: string[],
) => { [type: string]: number } = (
	nodeContent: (ADFEntity | undefined)[],
	allowedTypes: string[],
) => {
	const childrenTypes: { [key: string]: number } = {};
	nodeContent.forEach((childNode, index) => {
		if (!childNode?.type || !allowedTypes.includes(childNode.type)) {
			return;
		}

		if (!childrenTypes[childNode.type]) {
			childrenTypes[childNode.type] = 1;
			return;
		}

		childrenTypes[childNode.type]++;
	});

	return childrenTypes;
};

const removeDuplicatedNodes = (
	type: string,
	content: (ADFEntity | undefined)[],
	predicate: (node: ADFEntity | undefined) => boolean,
) => {
	let maxIterations = 10;
	let childrenTypeCounts = getChildrenTypeCounts(content, [type])[type];
	let firstPredicateNodeIndex = content.findIndex(predicate);
	while (childrenTypeCounts > 1 && firstPredicateNodeIndex > -1 && maxIterations-- > 0) {
		content.splice(firstPredicateNodeIndex, 1);
		firstPredicateNodeIndex = content.findIndex(predicate);
		childrenTypeCounts = getChildrenTypeCounts(content, [type])[type];
	}
};

/**
 * @param  {{[type:string]:number}} allowedTypes - array types allowed to deduplicate
 */
const deduplicateMediaSingleChildren = (
	node: ADFEntity,
	allowedTypes: { [type: string]: number },
) => {
	if (!node.content) {
		return;
	}

	const content = [...node.content];

	Object.keys(allowedTypes).forEach((type) => {
		//prioritise removing empty nodes first
		removeDuplicatedNodes(type, content, (node) => node?.type === type && isEmpty(node));
		//remove other remaining dupicates
		removeDuplicatedNodes(type, content, (node) => node?.type === type);
	});
	return {
		...node,
		content,
	};
};

export const transformInvalidMediaContent = (adf: ADFEntity) => {
	let isTransformed: boolean = false;
	const transformedAdf = traverse(adf, {
		mediaSingle: (node) => {
			if (!node?.content?.length) {
				return;
			}

			const disallowedDuplicateTypes = ['media', 'caption'];
			const childrenTypes = getChildrenTypeCounts(node.content, disallowedDuplicateTypes);
			if (Object.values(childrenTypes).some((occurences) => occurences > 1)) {
				isTransformed = true;
				return deduplicateMediaSingleChildren(node, childrenTypes);
			}
		},
	}) as ADFEntity;

	return {
		transformedAdf,
		isTransformed,
	};
};
