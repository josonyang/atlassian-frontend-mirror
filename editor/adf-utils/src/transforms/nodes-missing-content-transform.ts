import { traverse } from '../traverse/traverse';
import { type ADFEntity } from '../types';
import { tableRow, tableCell, paragraph, listItem } from '../builders';
import type { TableCellDefinition, TextDefinition } from '@atlaskit/adf-schema';
import { isEmpty } from './helpers';

const getMaxColumnsCountForTable = (tableNode: ADFEntity): number => {
	let colsInRow = 1;
	tableNode.content?.forEach((childNode) => {
		if (
			childNode?.type === 'tableRow' &&
			typeof childNode.content?.length === 'number' &&
			childNode.content.length > colsInRow
		) {
			colsInRow = childNode.content.length;
		}
	});
	return colsInRow;
};

const createValidEmptyContent = (node: ADFEntity): ADFEntity[] => {
	switch (node.type) {
		case 'tableCell':
			return [paragraph()];
		default:
			return [];
	}
};

const fixIfTableCellInvalidEmpty = (reportTransform: ReportTransform) => (node: ADFEntity) => {
	if (isEmpty(node)) {
		reportTransform();
		return {
			...node,
			content: createValidEmptyContent(node),
		};
	}
	return;
};

const hasNonListItemChildren = (node: ADFEntity) =>
	node.content?.some((node) => node?.type !== 'listItem');

const hasEmptyListItemChildren = (node: ADFEntity) =>
	node.content?.some((childNode) => childNode?.type === 'listItem' && !childNode.content?.length);

const tryCreateValidListItemWrappedChildren = (parentListNode: ADFEntity) => {
	return parentListNode.content?.map((childNode) => {
		if (childNode) {
			switch (childNode.type) {
				case 'listItem': {
					if (isEmpty(childNode)) {
						const result = listItem([paragraph()]);
						return result;
					}
					return childNode;
				}
				case 'text':
					return listItem([paragraph(childNode as TextDefinition)]);
				default:
					// Ignored via go/ees005
					// eslint-disable-next-line @typescript-eslint/no-explicit-any
					return listItem([childNode as any]);
			}
		}
		return childNode;
	});
};

const fixIfListParentWithInvalidListItemChildren =
	(reportTransform: ReportTransform) => (node: ADFEntity) => {
		if (hasNonListItemChildren(node) || hasEmptyListItemChildren(node)) {
			reportTransform();
			return {
				...node,
				content: tryCreateValidListItemWrappedChildren(node),
			};
		}
		return;
	};

const hasNonTableRowChildren = (node: ADFEntity) =>
	node.content?.some((node) => node?.type !== 'tableRow');

const tryCreateValidTableRowWrappedChildren = (parentTableNode: ADFEntity) => {
	const maxColsCount = getMaxColumnsCountForTable(parentTableNode);

	return parentTableNode.content?.map((childNode) => {
		if (childNode) {
			switch (childNode.type) {
				case 'text': {
					return tableRow([
						tableCell({})(paragraph(childNode as TextDefinition)),
						...new Array(maxColsCount - 1).fill(tableCell({})(paragraph())),
					]);
				}
				case 'tableCell': {
					return tableRow([childNode as TableCellDefinition]);
				}
				case 'tableRow': {
					if (isEmpty(childNode)) {
						return tableRow([...new Array(maxColsCount).fill(tableCell({})(paragraph()))]);
					}
					return childNode;
				}
				default:
					return childNode;
			}
		}
		return childNode;
	});
};

const hasEmptyTableRowChildren = (node: ADFEntity) =>
	node?.content?.some((node) => node?.type === 'tableRow' && node?.content?.length === 0) ||
	undefined;

const fixIfTableParentWithInvalidTableRowChildren =
	(reportTransform: ReportTransform) => (node: ADFEntity) => {
		if (hasEmptyTableRowChildren(node) || hasNonTableRowChildren(node)) {
			reportTransform();
			return {
				...node,
				content: tryCreateValidTableRowWrappedChildren(node),
			};
		}
		return;
	};

const removeMediaSingleWithNoContent = (reportTransform: ReportTransform) => (node: ADFEntity) => {
	if (isEmpty(node)) {
		reportTransform();
		return false;
	}
	return;
};

type ReportTransform = () => void;

interface TransformNodesMissingContentResult {
	isTransformed: boolean;
	transformedAdf: ADFEntity;
}

export const transformNodesMissingContent = (
	adf: ADFEntity,
): TransformNodesMissingContentResult => {
	let isTransformed: boolean = false;

	const reportTransform: ReportTransform = () => {
		isTransformed = true;
	};

	let transformedAdf = traverse(adf, {
		tableCell: fixIfTableCellInvalidEmpty(reportTransform),
	});

	// Empty mediaSingle nodes get stripped from the document and so this transform
	// will create an empty listItem if the media node is in a list. Empty listItems
	// are invalid adf, so we need to transform the mediaSingle nodes first.
	transformedAdf = traverse(transformedAdf || adf, {
		mediaSingle: removeMediaSingleWithNoContent(reportTransform),
	});

	transformedAdf = traverse(transformedAdf || adf, {
		bulletList: fixIfListParentWithInvalidListItemChildren(reportTransform),
		orderedList: fixIfListParentWithInvalidListItemChildren(reportTransform),
		table: fixIfTableParentWithInvalidTableRowChildren(reportTransform),
	}) as ADFEntity;

	return {
		transformedAdf,
		isTransformed,
	};
};
