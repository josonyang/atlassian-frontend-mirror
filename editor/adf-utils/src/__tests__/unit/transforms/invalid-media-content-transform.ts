import { transformInvalidMediaContent } from '../../../transforms';

import mediaSingleWithDuplicateCaptions1 from './__fixtures__/mediasingle-with-duplicate-captions-1.json';
import mediaSingleWithDuplicateCaptions2 from './__fixtures__/mediasingle-with-duplicate-captions-2.json';
import mediaSingleWithDuplicateCaptions3 from './__fixtures__/mediasingle-with-duplicate-captions-3.json';
import mediaSingleWithDuplicateMedia from './__fixtures__/mediasingle-with-duplicate-media.json';
import mediaSingleWithDuplicateMediaAndCaptions from './__fixtures__/mediasingle-with-duplicate-media-and-captions.json';

describe('transformInvalidMediaContent', () => {
	describe('mediaSingle', () => {
		it.each([
			['with duplicated captions and one empty', mediaSingleWithDuplicateCaptions1],
			['with duplicated nonempty captions', mediaSingleWithDuplicateCaptions2],
			['with duplicated media', mediaSingleWithDuplicateMedia],
			['with duplicated media and captions', mediaSingleWithDuplicateMediaAndCaptions],
			['with multiple empty and one non-empty caption', mediaSingleWithDuplicateCaptions3],
		])('%s', (_, adf) => {
			const { isTransformed, transformedAdf } = transformInvalidMediaContent(adf);
			expect(isTransformed).toEqual(true);
			expect(transformedAdf).toMatchSnapshot();
		});
	});
});
