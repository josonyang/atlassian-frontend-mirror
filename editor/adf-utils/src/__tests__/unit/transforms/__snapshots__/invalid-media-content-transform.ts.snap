// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformInvalidMediaContent mediaSingle with duplicated captions and one empty 1`] = `
{
  "content": [
    {
      "attrs": {
        "layout": "center",
        "width": 33.33333333333333,
      },
      "content": [
        {
          "attrs": {
            "collection": "contentId-0000000000",
            "height": 400,
            "id": "00000000-0000-0000-0000-0000000000",
            "occurrenceKey": null,
          },
          "type": "media",
        },
        {
          "content": [
            {
              "text": "caption 1",
              "type": "text",
            },
          ],
          "type": "caption",
        },
      ],
      "type": "mediaSingle",
    },
  ],
  "type": "doc",
}
`;

exports[`transformInvalidMediaContent mediaSingle with duplicated media 1`] = `
{
  "content": [
    {
      "attrs": {
        "layout": "center",
        "width": 33.33333333333333,
      },
      "content": [
        {
          "attrs": {
            "collection": "contentId-0000000001",
            "height": 400,
            "id": "00000000-0000-0000-0000-0000000001",
            "occurrenceKey": null,
          },
          "type": "media",
        },
      ],
      "type": "mediaSingle",
    },
  ],
  "type": "doc",
}
`;

exports[`transformInvalidMediaContent mediaSingle with duplicated media and captions 1`] = `
{
  "content": [
    {
      "attrs": {
        "layout": "center",
        "width": 33.33333333333333,
      },
      "content": [
        {
          "attrs": {
            "collection": "contentId-0000000001",
            "height": 400,
            "id": "00000000-0000-0000-0000-0000000001",
            "occurrenceKey": null,
          },
          "type": "media",
        },
        {
          "content": [
            {
              "text": "caption 2",
              "type": "text",
            },
          ],
          "type": "caption",
        },
      ],
      "type": "mediaSingle",
    },
  ],
  "type": "doc",
}
`;

exports[`transformInvalidMediaContent mediaSingle with duplicated nonempty captions 1`] = `
{
  "content": [
    {
      "attrs": {
        "layout": "center",
        "width": 33.33333333333333,
      },
      "content": [
        {
          "attrs": {
            "collection": "contentId-0000000000",
            "height": 400,
            "id": "00000000-0000-0000-0000-0000000000",
            "occurrenceKey": null,
          },
          "type": "media",
        },
        {
          "content": [
            {
              "text": "caption 2",
              "type": "text",
            },
          ],
          "type": "caption",
        },
      ],
      "type": "mediaSingle",
    },
  ],
  "type": "doc",
}
`;

exports[`transformInvalidMediaContent mediaSingle with multiple empty and one non-empty caption 1`] = `
{
  "content": [
    {
      "attrs": {
        "layout": "center",
        "width": 33.33333333333333,
      },
      "content": [
        {
          "attrs": {
            "collection": "contentId-0000000000",
            "height": 400,
            "id": "00000000-0000-0000-0000-0000000000",
            "occurrenceKey": null,
          },
          "type": "media",
        },
        {
          "content": [
            {
              "text": "caption 1",
              "type": "text",
            },
          ],
          "type": "caption",
        },
      ],
      "type": "mediaSingle",
    },
  ],
  "type": "doc",
}
`;
